pkg/reconciler/revision/resources/deploy_test.go

```diff
@@ -680,28 +902,99 @@ func TestMakePodSpec(t *testing.T) {
 		),
 		want: podSpec(
 			[]corev1.Container{
-				userContainer(
+				servingContainer(
+					func(container *corev1.Container) {
+						container.Image = "busybox@sha256:deadbeef"
+					},
+					withContainer(),
+					withEnvVar("K_CONFIGURATION", "cfg"),
+					withEnvVar("K_SERVICE", "svc"),
+				),
+				queueContainer(
+					withEnvVar("CONTAINER_CONCURRENCY", "1"),
+					withEnvVar("SERVING_SERVICE", "svc"),
+				),
+			}),
+	}, {
+		name: "complex pod spec for multiple containers with container data to all containers",
+		rev: revision("bar", "foo",
+			withContainers([]corev1.Container{
+				{
+					Name:  servingContainerName,
+					Image: "busybox",
+					Ports: []corev1.ContainerPort{{
+						ContainerPort: 8080,
+					}},
+				}, {
+					Name:  sidecarContainerName,
+					Image: "ubuntu",
+				}, {
+					Name:  "sidecar-container-2",
+					Image: "alpine",
+				},
+			}),
+			withContainerConcurrency(1),
+			WithContainerStatuses([]v1.ContainerStatuses{{
+				ImageDigest: "busybox@sha256:deadbeef",
+			}, {
+				ImageDigest: "ubuntu@sha256:deadbffe",
+			}, {
+				ImageDigest: "alpine@sha256:deadbfff",
+			}}),
+			func(revision *v1.Revision) {
+				revision.ObjectMeta.Labels = map[string]string{
+					serving.ConfigurationLabelKey: "cfg",
+					serving.ServiceLabelKey:       "svc",
+				}
+				// When there are multiple containers then only serving container will have readinessProbe. here
+				// container[0] acts as a serving container because of the provided containerPort so using withTCPReadinessProbe.
+				container(&revision.Spec.Containers[0],
+					withTCPReadinessProbe(),
+				)
+				for i := range revision.Spec.Containers {
+					revision.Spec.Containers[i].Command = []string{"/bin/bash"}
+					revision.Spec.Containers[i].Args = []string{"-c", "echo Hello world"}
+					container(&revision.Spec.Containers[i],
+						withEnvVar("FOO", "bar"),
+						withEnvVar("BAZ", "blah"),
+					)
+					revision.Spec.Containers[i].Resources = corev1.ResourceRequirements{
+						Requests: corev1.ResourceList{
+							corev1.ResourceMemory: resource.MustParse("666Mi"),
+							corev1.ResourceCPU:    resource.MustParse("666m"),
+						},
+						Limits: corev1.ResourceList{
+							corev1.ResourceMemory: resource.MustParse("888Mi"),
+							corev1.ResourceCPU:    resource.MustParse("888m"),
+						},
+					}
+					revision.Spec.Containers[i].TerminationMessagePolicy = corev1.TerminationMessageReadFile
```

Comment:
I'd rather have all of these in the container definitions above. This is hard to compile in-head to figure out what actually happens.

Reply:
All these informations are not part of all the testcase so added separate function `updateContainer` to update the existing container to these values and these informations are common in 3 testcases